client/sections-middleware.js

```diff
@@ -0,0 +1,112 @@
+import { switchCSS } from 'lib/i18n-utils/switch-locale';
+
+import sections from './sections';
+
+const _loadedSections = {};
+
+function maybeLoadCSS( sectionName ) {
+	//eslint-disable-line no-unused-vars
+	const section = find( sections, { name: sectionName } );
+
+	if ( ! ( section && section.css ) ) {
+		return;
+	}
+
+	const url =
+		typeof document !== 'undefined' && document.documentElement.dir === 'rtl'
+			? section.css.urls.rtl
+			: section.css.urls.ltr;
+
+	switchCSS( 'section-css-' + section.css.id, url );
+}
+
+function preload( sectionName ) {
+	maybeLoadCSS( sectionName );
+	const filteredSections = filter( sections, { name: sectionName } );
+	if ( isEmpty( sections ) ) {
+		return Promise.reject( `Attempting to load non-existent section: ${ sectionName }` );
+	}
+	return Promise.all( filteredSections.map( section => section.load() ) );
+}
+
+preloadHub.on( 'preload', preload );
+
+function activateSection( sectionDefinition, context, next ) {
+	const dispatch = context.store.dispatch;
+
+	controller.setSection( sectionDefinition )( context );
+	dispatch( { type: 'SECTION_SET', isLoading: false } );
+	dispatch( activateNextLayoutFocus() );
+	next();
+}
+
+function createPageDefinition( path, sectionDefinition ) {
+	const pathRegex = pathToRegExp( path );
+
+	page( pathRegex, function( context, next ) {
+		const envId = sectionDefinition.envId;
+		const dispatch = context.store.dispatch;
+
+		if ( envId && envId.indexOf( config( 'env_id' ) ) === -1 ) {
+			return next();
+		}
+		if ( _loadedSections[ sectionDefinition.module ] ) {
+			return activateSection( sectionDefinition, context, next );
+		}
+		if ( config.isEnabled( 'restore-last-location' ) && restoreLastSession( context.path ) ) {
+			return;
+		}
+		dispatch( { type: 'SECTION_SET', isLoading: true } );
+		preload( sectionDefinition.name )
+			.then( requiredModules => {
+				if ( ! _loadedSections[ sectionDefinition.module ] ) {
+					requiredModules.forEach( mod => mod( controller.clientRouter ) ); // if we do array
+					_loadedSections[ sectionDefinition.module ] = true;
+				}
+				return activateSection( sectionDefinition, context, next );
+			} )
+			.catch( error => {
+				console.error( error ); // eslint-disable-line
+				if ( ! LoadingError.isRetry() ) {
+					LoadingError.retry( sectionDefinition.name );
+				} else {
+					dispatch( { type: 'SECTION_SET', isLoading: false } );
+					LoadingError.show( sectionDefinition.name );
+				}
+			} );
+	} );
+}
+
+export const getSections = () => sections;
+
+export const setupRoutes = () => {
+	sections.forEach( section =>
+		section.paths.forEach( path => createPageDefinition( path, section ) )
+	);
+};
+
+// Adapts route paths to also include wildcard
+// subroutes under the root level section.
+export function pathToRegExp( path ) {
+	// Prevents root level double dash urls from being validated.
+	if ( path === '/' ) {
+		return path;
+	}
+	return new RegExp( '^' + path + '(/.*)?$' );
```

Comment:
Lets [escape](https://github.com/Automattic/wp-calypso/pull/21255/files#diff-b2907500f646f51484f42b858d1d63b6R8) `path` with `escapeRegExp`

Reply:
Isn't `escapeRegExp` redundant, given https://github.com/Automattic/wp-calypso/pull/21349/files#diff-afb3c8d4236d194988c31a97e465126dR108 - 110?

On a sidenote, IIRC we're using the RegEx here so that a section definition like https://github.com/Automattic/wp-calypso/blob/abb552dfba238d93315629bebb4939bcaaaf5c6a/client/wordpress-com.js#L205 matches both `/theme` and `/theme/*`. However, a RegEx seems almost overkill to do this, when `page.js` also comes with its ownpseudo-regular syntax. Maybe we should just map each string in a section definition's `paths` array into two -- one unchanged, and one with `/*` appended.